csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,191 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <disjoint_set.h>
+#include <ir/all_nodes.h>
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace nvfuser {
+
+using IdGroup = std::shared_ptr<VectorOfUniqueEntries<IterDomain*>>;
+using IdGroups = VectorOfUniqueEntries<IdGroup>;
+using ExprGroup = std::shared_ptr<VectorOfUniqueEntries<Expr*>>;
+using ExprGroups = VectorOfUniqueEntries<ExprGroup>;
```

Comment:
Could you add a comment providing an example on what these classes are? For example, if I have a fusion:

```C++
T1 = set(T0);
T2 = set(T1);

T0: root [I0, I1], leaf [I0, I1]
T1: root [I2, I3], leaf [I2*I3/4, 4]
T2: root [I4, I5], leaf [I4*I5/4, 4]
```
What will the exact map looks like? What `IdGroup` and what `IdGroups` will we have? And what `ExprGroup` and what `ExprGroups` will we have?

Reply:
In this case, there would be `IdGroup`s of:

```
- {I0, I2, I4}
- {I1, I3, I5}
- {I2*I3, I4*I5}
- {I2*I3/4, I4*I5/4}
- {4, 4}
```

There would be `ExprGroup`s of:

```
- {merge of I2 and I3, merge of I4 and I5}
- {split of I2*I3, split of I4*I5}
```

`IdGroups` and `ExprGroups` are just used when there are multiple of `IdGroup`s and `ExprGroup`s.

Both `IdGroup` and `ExprGroup` are the same type as the group type of `DisjointSets<IterDomain*>` and `DisjointSets<Expr*>`, respectively.

Reply:
Could you add this as a comment in the code? If you think https://github.com/NVIDIA/Fuser/pull/1168#issuecomment-1789714911 makes sense, the example there could also serve as an example in the code.

Reply:
Added